editor/src/document/document_message_handler.rs

```diff
@@ -132,55 +155,54 @@ impl MessageHandler<DocumentsMessage, &InputPreprocessor> for DocumentsMessageHa
 			CloseAllDocuments => {
 				// Empty the list of internal document data
 				self.documents.clear();
+				self.document_ids.clear();
 
 				// Create a new blank document
 				responses.push_back(NewDocument.into());
 			}
-			CloseDocument(id) => {
-				assert!(id < self.documents.len(), "Tried to select a document that was not initialized");
-				// Remove doc from the backend store; use `id` as client tabs and backend documents will be in sync
-				self.documents.remove(id);
+			CloseDocument(index) => {
+				assert!(index < self.documents.len(), "Tried to select a document that was not initialized");
+				// Get the ID based on the current collection of the documents.
+				let id = self.document_ids[index];
+				// Map the ID to an index and remove the document
+				self.documents.remove(&id);
+				self.document_ids.remove(index);
 
-				// Send the new list of document tab names
-				let open_documents = self.documents.iter().map(|doc| doc.name.clone()).collect();
-				responses.push_back(FrontendMessage::UpdateOpenDocumentsList { open_documents }.into());
+				self.active_document_index = if self.active_document_index >= self.document_ids.len() {
+					let len = self.document_ids.len();
+					(self.active_document_index + len - 1) % len
+				} else {
+					index
+				};
 
-				// Last tab was closed, so create a new blank tab
-				if self.documents.is_empty() {
-					self.active_document_index = 0;
-					responses.push_back(NewDocument.into());
+				// // Last tab was closed, so create a new blank tab
+				if self.document_ids.is_empty() {
+					self.counter = self.counter + 1;
+					self.document_ids.push(self.counter);
+					self.documents.insert(self.counter, DocumentMessageHandler::default());
 				}
-				// The currently selected doc is being closed
-				else if id == self.active_document_index {
-					// The currently selected tab was the rightmost tab
-					if id == self.documents.len() {
-						self.active_document_index -= 1;
-					}
 
-					responses.push_back(DocumentMessage::DocumentStructureChanged.into());
-					responses.push_back(
-						FrontendMessage::SetActiveDocument {
-							document_index: self.active_document_index,
-						}
-						.into(),
-					);
-					responses.push_back(
-						FrontendMessage::UpdateCanvas {
-							document: self.active_document_mut().graphene_document.render_root(),
-						}
-						.into(),
-					);
-				}
-				// Active doc will move one space to the left
-				else if id < self.active_document_index {
-					self.active_document_index -= 1;
-					responses.push_back(
-						FrontendMessage::SetActiveDocument {
-							document_index: self.active_document_index,
+				// Send the new list of document tab names
+				let open_documents: Vec<String> = self
```

Comment:
just fyi you don't need type annotations here, `Vec<_>` is enough but you can add them for clearity